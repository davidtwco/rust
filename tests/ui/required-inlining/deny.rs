//@ build-fail
//@ compile-flags: --crate-type=lib
#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(required_inlining)]

// Test that required inlining w/ errors works as expected.

#[rustc_no_mir_inline]
#[inline(required)]
pub fn callee() {
}

#[rustc_no_mir_inline]
#[inline(required("maintain security properties"))]
pub fn callee_justified() {
}

pub fn caller() {
    callee();
//~^ ERROR `callee` could not be inlined into `caller` but is required to be inlined

    callee_justified();
//~^ ERROR `callee_justified` could not be inlined into `caller` but is required to be inlined
}

#[warn(required_inline)]
pub fn caller_overridden() {
    callee();
//~^ WARN `callee` could not be inlined into `caller_overridden` but is required to be inlined

    callee_justified();
//~^ WARN `callee_justified` could not be inlined into `caller_overridden` but is required to be inlined
}
