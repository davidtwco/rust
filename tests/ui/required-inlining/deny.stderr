error: `callee` could not be inlined into `caller` but is required to be inlined
  --> $DIR/deny.rs:20:5
   |
LL | #[inline(required)]
   | ------------------- inlining due to this annotation
LL | pub fn callee() {
   | --------------- `callee` defined here
...
LL | pub fn caller() {
   | --------------- within `caller`...
LL |     callee();
   |     ^^^^^^^^ ...`callee` called here
   |
   = note: could not be inlined due to: #[rustc_no_mir_inline]
   = note: `#[deny(required_inline)]` on by default

error: `callee_justified` could not be inlined into `caller` but is required to be inlined
  --> $DIR/deny.rs:23:5
   |
LL | #[inline(required("maintain security properties"))]
   | --------------------------------------------------- inlining due to this annotation
LL | pub fn callee_justified() {
   | ------------------------- `callee_justified` defined here
...
LL | pub fn caller() {
   | --------------- within `caller`...
...
LL |     callee_justified();
   |     ^^^^^^^^^^^^^^^^^^ ...`callee_justified` called here
   |
   = note: could not be inlined due to: #[rustc_no_mir_inline]
   = note: `callee_justified` is required to be inlined to: maintain security properties

warning: `callee` could not be inlined into `caller_overridden` but is required to be inlined
  --> $DIR/deny.rs:29:5
   |
LL | #[inline(required)]
   | ------------------- inlining due to this annotation
LL | pub fn callee() {
   | --------------- `callee` defined here
...
LL | pub fn caller_overridden() {
   | -------------------------- within `caller_overridden`...
LL |     callee();
   |     ^^^^^^^^ ...`callee` called here
   |
   = note: could not be inlined due to: #[rustc_no_mir_inline]
note: the lint level is defined here
  --> $DIR/deny.rs:27:8
   |
LL | #[warn(required_inline)]
   |        ^^^^^^^^^^^^^^^

warning: `callee_justified` could not be inlined into `caller_overridden` but is required to be inlined
  --> $DIR/deny.rs:32:5
   |
LL | #[inline(required("maintain security properties"))]
   | --------------------------------------------------- inlining due to this annotation
LL | pub fn callee_justified() {
   | ------------------------- `callee_justified` defined here
...
LL | pub fn caller_overridden() {
   | -------------------------- within `caller_overridden`...
...
LL |     callee_justified();
   |     ^^^^^^^^^^^^^^^^^^ ...`callee_justified` called here
   |
   = note: could not be inlined due to: #[rustc_no_mir_inline]
   = note: `callee_justified` is required to be inlined to: maintain security properties

error: aborting due to 2 previous errors; 2 warnings emitted

